core/src/main/java/com/linecorp/armeria/common/util/DefaultBlockingTaskExecutor.java

```diff
@@ -83,50 +154,163 @@ public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedE
 
     @Override
     public <T> Future<T> submit(Callable<T> task) {
-        return delegate.submit(task);
+        boolean submitted = false;
+        taskCounter.incrementAndGet();
+        try {
+            final Future<T> future = delegate.submit(() -> {
+                try {
+                    return task.call();
+                } finally {
+                    taskCounter.decrementAndGet();
+                }
+            });
+            submitted = true;
+            return future;
+        } finally {
+            if (!submitted) {
+                taskCounter.decrementAndGet();
+            }
+        }
     }
 
     @Override
     public <T> Future<T> submit(Runnable task, T result) {
-        return delegate.submit(task, result);
+        boolean submitted = false;
+        taskCounter.incrementAndGet();
+        try {
+            final Future<T> future = delegate.submit(() -> {
+                try {
+                    task.run();
+                } finally {
+                    taskCounter.decrementAndGet();
+                }
+            }, result);
+            submitted = true;
+            return future;
+        } finally {
+            if (!submitted) {
+                taskCounter.decrementAndGet();
+            }
+        }
     }
 
     @Override
     public Future<?> submit(Runnable task) {
-        return delegate.submit(task);
+        boolean submitted = false;
+        taskCounter.incrementAndGet();
+        try {
+            final Future<?> future = delegate.submit(() -> {
+                try {
+                    task.run();
+                } finally {
+                    taskCounter.decrementAndGet();
+                }
+            });
+            submitted = true;
+            return future;
```

Comment:
If `future.cancel` is called immediately without given a chance to run, I think it is possible that `taskCounter` is incremented only without decrementing.

I think we may also need to wrap the returned future and check if `cancel` is called successfully.

Reply:
https://github.com/line/armeria/pull/4073#issuecomment-1066526496

Nevermind, I saw this comment late.

We may need something like this then:


```java
  class BlockingFuture<T> implements ScheduledFuture<T> {
    private final Future<T> delegate;
    BlockingFuture(Future<T> delegate) {
        this.delegate = delegate;
    }
    @Override
    public boolean cancel(boolean mayInterruptIfRunning) {
        boolean cancelled = delegate.cancel(mayInterruptIfRunning);
        if (cancelled) {
            taskCounter.decrementAndGet();
        }
        return cancelled;
    }
```